### DISCUSSION TITLE
Slow LTSP fat client boot times compared to old LTSP Version

### DISCUSSION THREAD

**Discussion by X-Dimension-88MPH** [ltsp/ltsp#816] [Q&A] [answered]
Hi! 
We are using some old Ubuntu 16.04 LTSP server in our schools and want to migrate to Ubuntu 22.04 with the new LTSP version.
I have installed two virtual machines: one with our old Ubuntu 16.04 LTSP system and one with Ubuntu 22.04.
Although Gnome runs very fast on fat clients using the new Ubuntu 22.04 based system, the client boot times are extremly slow compared to the Ubuntu 16.04 LTSP setup.

Ubuntu 16.04
- DHCP offer - 5 seconds 
- PXE - 5 seconds
- OS  -  5 seconds
Total: 15 seconds from power on to login manager 

Ubuntu 22.04
- DHCP offer - 5 seconds 
- PXE - 25 seconds
- OS boot - 10 seconds
Total: 40 seconds from power on to login manager 

It looks like the PXE part takes the longest time. Starting iPXE and loading the Client image from the Server seems to be much slower than on the old LTSP version. 

The only differences between my Ubuntu 16.04 and 22.04 LTSP setups are that Ubuntu 16.04 is configured with chroot client method and Ubuntu 22.04 with chrootless (previously pnp) .

Is it possible that the chroot vs chrootless setup is the reason for the slower boot on the new LTSP Server, or is something wrong with my PXE setting? (i disabled only PXE boot menu, all other settings are LTSP defaults) 

Thanks for help. :)

**Comment by X-Dimension-88MPH** [comment]
BTW:
I can reduce the startup time from 40 to 30 seconds when our fat clients are booting this ipxe.iso from a usb drive.  
https://boot.ipxe.org/ipxe.iso

**Comment by alkisg** [comment, marked as answer]
I think there are two parameters involved:

- The kernel and initrd size is a lot bigger in 22.04 compared with 16.04
- The new ipxe is also slower in your setup compared to ipxe.iso or to the old one

To verify that, I would records two videos, one where ipxe is netbooted and one where ipxe.iso is used.
The videos will make it easier to convince others that ipxe is to blame.
Then I would contact the ipxe guys and ask for advice on how to make things faster.
While there's also this test that you could do: https://ltsp.org/guides/pxelinux/

**Comment by X-Dimension-88MPH** [comment]
Here are some screen recordings:

1. Ubuntu 16.04 LTSP reboot
[Ubuntu-16.04-LTSP-boot.webm](https://user-images.githubusercontent.com/117190661/230366239-ff426461-4502-40ec-942c-a1dfe98e9bf6.webm)

2. Ubuntu 22.04 LTSP reboot
[Ubuntu-22.04-LTSP-boot.webm](https://user-images.githubusercontent.com/117190661/230366236-d1b3941b-2728-4698-bfad-c56f535153b5.webm)

3. Ubuntu 22.04 LTSP with ipxe.iso reboot
[Ubuntu-22.04-LTSP-ipxe-iso-boot.webm](https://user-images.githubusercontent.com/117190661/230366241-d97e6de7-9f38-4daa-8bd9-d9f76b057751.webm)

I will try your pxelinux guide and give you a feedback.

Thank you very much Alkis  :)

**Comment by alkisg** [comment]
OK, it looks fine to me. The kernel/initrd are bigger, and undionly.kpxe needs to re-initialize the adapter, while ipxe.pxe initializes it only once. Everything seems normal.

Indeed pxelinux will be a bit faster as it doesn't need to re-initialize the adapter, so if you care about these few seconds you can use that.

**Comment by X-Dimension-88MPH** [comment]
Your pxelinux guide seems to be the best solution!

Downloading and using ipxe.pxe instead of undionly.kpxe reduces boot time also by 10 seconds, so there is no need to boot from the ISO anymore. :+1: 

Using pxelinux instead ipxe looks also very promising!
But after installing pxelinux and syslinux i get a menu on boot like this:

PXE
     --> pxelinux.0
           undionly.kpxe

How can i disable this menu, so that pxelinux.0 is starting directly on boot?

Is it possible to hide all the pxe output of pxelinux or ipxe on boot, or reduce it a little bit?
Although it is helpful for testing purposes and finding errors, it's not very beautiful in a production environment. :see_no_evil:

**Comment by alkisg** [comment]
You can delete the line that mentions undionly.kpxe in /etc/dnsmasq.d/ltsp-dnsmasq.conf, and restart dnsmasq.

**Comment by X-Dimension-88MPH** [comment]
Works perfect! Thank you! 👍

**Comment by X-Dimension-88MPH** [comment]
Now I have some clients that only support UEFI and are switching to iPXE for booting.
I run into the same problem you have described here with Realtek onboard LAN adapter: https://github.com/ipxe/ipxe/issues/145

Since the last post there is over a year old and i see no solution, did you found a workaround for your problem @alkisg ?

My idea was to use syslinux.efi instead of iPXE for UEFI based computer, but i don't understand how to modify dnsmasq configuration to get it to work. 
At this point of my tests only syslinux.efi is loaded via PXE, but it seems to stop after that when it should load initrd.img. 
I can't see what's going on after syslinux loads because the client shows "no boot device" before i could read the last output of syslinux.

**Comment by alkisg** [comment]
I've seen that issue about 50 times now.
In about 45 of these, the problem was actually a bad connection somewhere and it was solved by replacing the cable or by fixing the wall network socket. So that ipxe bug actually helped us fix networking issues.
In some of the rest cases, the problem was the combination of gigabit networking with 100 mbps networking.
In only a couple of the cases like the one I describe in the bug, the problem appeared in full gigabit connections. So it was so rare that I didn't take the time to troubleshoot it more.

So first check the cables, and if necessary, try with a direct cable connection between the client and the server.
If that one is slow, then make sure that after the kernel is loaded, everything is fast again (i.e. after vmlinuz/initrd).
If that's the case, then you've hit the rare bug that I filed.

**Comment by X-Dimension-88MPH** [comment]
After some testing i can't find the reason why PXE Boot is so extremely slow on all our systems. The virtual machine i've used at the beginning is still the fastest one with 30-40 seconds total boot time. But this VM runs on the same physical host as our LTSP and  is directly connected to it via a virtual 10 GBit network card. Also the TFTP download speed is only at 60 MBit/s 

With all notebooks and computers i have tried the total boot times are between 110 seconds (Intel Core i3-4130 with 1GBit Intel network card)  and 145 seconds (AMD Ryzen 5 5650 with 1GBit Realtek network card)  The TFTP download speed lies between 30 and 40 MBit/s.

Is there a way to test if i have a problem with a slow connection speed or with slow TFTP?
Also i have found some tutorials that IPXE and PXELINUX support http for faster download speeds, but after some hours playing around with lpxelinux, syslinux.efi and ipxe, all systems only boot into a busybox and i don't know whats going wrong.

If anybody has fast LTSP client boot times and knows the reason of my problems i would be very thankful for a little help. :(

**Comment by alkisg** [comment]
I had the same problem yesterday. Then I changed the 40G cable with another one and the initrd was fetched in 30 seconds instead of 15 minutes.

I don't know exactly why TFTP/UDP is slow in that case, while TCP isn't; it might be related to flow control.
But so far in the vast majority of the cases the root cause has been a cabling problem for me.
And since I'm not sure what other UDP traffic will be impacted, I prefer to solve the root cause than find workarounds (e.g. http).

For http-only boot, there are discussions/issues/pull-requests here that show how to modify dnsmasq.conf and ltsp.ipxe to do it.

**Comment by X-Dimension-88MPH** [comment]
Again, many thanks @alkisg !  :+1: 

I'm now using lighttpd with its web root pointing to /srv/tftp and these lines you have posted in a pull request in my dnsmasq config:

pxe-service=tag:proxy,tag:iPXE,X86PC,"ltsp.ipxe","http://${proxydhcp/dhcp-server}/ltsp/ltsp.ipxe"
pxe-service=tag:proxy,tag:iPXE,X86-64_EFI,"ltsp.ipxe","http://${proxydhcp/dhcp-server}/ltsp/ltsp.ipxe"

This combination gives me a huge improvement and reduces boot times between 10 seconds on the already fast starting virtual machine and 50 seconds on the AMD Ryzen 5 based system! :sunglasses: 

Now the boot performance is comparable with our old Ubuntu 16.04 LTSP Server and the system is production ready for our schools.
